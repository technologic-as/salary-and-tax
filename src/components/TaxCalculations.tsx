import Grid from '@material-ui/core/Grid';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableHead from '@material-ui/core/TableHead';
import React, { Fragment } from 'react';
import { defineMessages, useIntl } from 'react-intl';
import { connect } from 'react-redux';
import { Section, TaxRow } from './Ui';
import { State } from '../reducers';

const messages = defineMessages({
  header: { id: 'tax.calculations.header', defaultMessage: 'Tax calculations' },
  income: { id: 'tax.calculations.income', defaultMessage: 'Income' },
  incomeHeader: {
    id: 'tax.calculations.income.header',
    defaultMessage: 'Income',
  },
  minimumDeduction: {
    id: 'tax.calculations.minimum.deduction',
    defaultMessage: 'Minimum deduction',
  },
  personalAllowance: {
    id: 'tax.calculations.personal.allowance',
    defaultMessage: 'Personal allowance',
  },
  commonIncome: {
    id: 'tax.calculations.common.income',
    defaultMessage: 'Common income',
  },
  commonIncomeHeader: {
    id: 'tax.calculations.common.income.header',
    defaultMessage: 'Common income',
  },
  surplusHeader: {
    id: 'tax.calculations.surplus.header',
    defaultMessage: 'Surplus',
  },
  surplus: {
    id: 'tax.calculations.surplus.amount',
    defaultMessage: 'Surplus amount',
  },
  surplusTax: {
    id: 'tax.calculations.surplus.tax',
    defaultMessage: 'Surplus tax',
  },
  dividends: {
    id: 'tax.calculations.dividends.amount',
    defaultMessage: 'Dividends',
  },
  dividendsTax: {
    id: 'tax.calculations.dividends.tax',
    defaultMessage: 'Dividends tax',
  },
  incomeTax: {
    id: 'tax.calculations.income.tax',
    defaultMessage: 'Income tax',
  },
  socialSecurityDeduction: {
    id: 'tax.calculations.social.security.deduction',
    defaultMessage: 'Social security deduction',
  },
  step1: { id: 'tax.calculations.step.1', defaultMessage: 'Tax step 1' },
  step2: { id: 'tax.calculations.step.2', defaultMessage: 'Tax step 2' },
  step3: { id: 'tax.calculations.step.3', defaultMessage: 'Tax step 3' },
  step4: { id: 'tax.calculations.step.4', defaultMessage: 'Tax step 4' },
  afterTax: { id: 'tax.calculations.after.tax', defaultMessage: 'After tax' },
});

type Props = {
  income: number;
  minimumDeduction: number;
  personalAllowance: number;
  commonIncome: number;
  incomeTax: number;
  socialSecurityDeduction: number;
  step1: number;
  step2: number;
  step3: number;
  step4: number;
  totalTax: number;
  afterTax: number;
  surplus: number;
  surplusTax: number;
  dividends: number;
  dividendsTax: number;
  afterDividendsTax: number;
};

export const TaxCalculationsComponent = ({
  income,
  minimumDeduction,
  personalAllowance,
  commonIncome,
  incomeTax,
  socialSecurityDeduction,
  step1,
  step2,
  step3,
  step4,
  totalTax,
  afterTax,
  surplus,
  dividends,
  dividendsTax,
  afterDividendsTax,
  surplusTax,
}: Props) => {
  const { formatMessage, formatNumber } = useIntl();
  const formatCurrency = (amount: number) => formatNumber(amount, { style: 'currency', currency: 'NOK' });
  return (
    <Fragment>
      <Section header={formatMessage(messages.header)} expanded>
        <Grid container>
          <Grid item>
            <Table>
              <TableHead>
                <TaxRow description={formatMessage(messages.commonIncomeHeader)} />
              </TableHead>
              <TableBody>
                <TaxRow description={formatMessage(messages.income)} sum={formatCurrency(income)} />
                <TaxRow
                  description={formatMessage(messages.minimumDeduction)}
                  amount={formatCurrency(minimumDeduction)}
                  minus
                />
                <TaxRow
                  description={formatMessage(messages.personalAllowance)}
                  amount={formatCurrency(personalAllowance)}
                  minus
                />
                <TaxRow description={formatMessage(messages.commonIncome)} sum={formatCurrency(commonIncome)} />
              </TableBody>
            </Table>
          </Grid>
          <Grid item>
            <Table>
              <TableHead>
                <TaxRow description={formatMessage(messages.incomeHeader)} />
              </TableHead>
              <TableBody>
                <TaxRow description={formatMessage(messages.income)} sum={formatCurrency(income)} />
                <TaxRow description={formatMessage(messages.incomeTax)} amount={formatCurrency(incomeTax)} minus />
                <TaxRow
                  description={formatMessage(messages.socialSecurityDeduction)}
                  amount={formatCurrency(socialSecurityDeduction)}
                  minus
                />
                <TaxRow description={formatMessage(messages.step1)} amount={formatCurrency(step1)} minus />
                <TaxRow description={formatMessage(messages.step2)} amount={formatCurrency(step2)} minus />
                <TaxRow description={formatMessage(messages.step3)} amount={formatCurrency(step3)} minus />
                <TaxRow description={formatMessage(messages.step4)} amount={formatCurrency(step4)} minus />
                <TaxRow
                  description={formatMessage(messages.afterTax)}
                  amount={formatCurrency(totalTax)}
                  sum={formatCurrency(afterTax)}
                />
              </TableBody>
            </Table>
          </Grid>
          <Grid item>
            <Table>
              <TableHead>
                <TaxRow description={formatMessage(messages.surplusHeader)} />
              </TableHead>
              <TableBody>
                <TaxRow description={formatMessage(messages.surplus)} sum={formatCurrency(surplus)} />
                <TaxRow
                  description={formatMessage(messages.surplusTax)}
                  amount={formatCurrency(surplusTax)}
                  sum={formatCurrency(dividends)}
                  minus
                />
                <TaxRow
                  description={formatMessage(messages.dividendsTax)}
                  amount={formatCurrency(dividendsTax)}
                  sum={formatCurrency(afterDividendsTax)}
                  minus
                />
                <TaxRow description={formatMessage(messages.dividends)} sum={formatCurrency(afterDividendsTax)} />
              </TableBody>
            </Table>
          </Grid>
        </Grid>
      </Section>
    </Fragment>
  );
};

const mapStateToProps = ({
  calculations: {
    tax: {
      income,
      minimumDeduction,
      personalAllowance,
      commonIncome,
      incomeTax,
      socialSecurityDeduction,
      step1,
      step2,
      step3,
      step4,
      totalTax,
      afterTax,
    },
    dividends: { surplus, surplusTax, dividendsTax, afterDividendsTax, dividends },
  },
}: State) => ({
  income,
  minimumDeduction,
  personalAllowance,
  commonIncome,
  incomeTax,
  socialSecurityDeduction,
  step1,
  step2,
  step3,
  step4,
  totalTax,
  afterTax,
  surplus,
  surplusTax,
  dividends,
  dividendsTax,
  afterDividendsTax,
});

const mapDispatchToProps = () => ({});

export const TaxCalculations = connect(mapStateToProps, mapDispatchToProps)(TaxCalculationsComponent);

export default TaxCalculations;
