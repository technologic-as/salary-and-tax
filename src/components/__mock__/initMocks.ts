const mockFormatMessage = ({ defaultMessage }: { defaultMessage: string }) => defaultMessage;
const mockFormatNumber = (amount: number) =>
  new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(amount);

const mockIntl = () => ({
  formatMessage: mockFormatMessage,
  formatDate: jest.fn,
  formatTime: jest.fn,
  formatRelative: jest.fn,
  formatNumber: mockFormatNumber,
  formatPlural: jest.fn,
  formatHTMLMessage: jest.fn,
  now: jest.fn,
});

jest.mock('react-intl', () => ({
  useIntl: mockIntl,
  defineMessages: (msg: unknown) => msg,
}));

jest.mock('react-intl-redux', () => ({
  updateIntl: (a: unknown) => a,
}));

jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');
jest.mock('@material-ui/core/FormGroup', () => 'FormGroup');
jest.mock('@material-ui/core/FormControlLabel', () => 'FormControlLabel');
jest.mock('@material-ui/core/Radio', () => 'Radio');

export {};
