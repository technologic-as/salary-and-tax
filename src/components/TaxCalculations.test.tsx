import mockComponent from 'identity-obj-proxy';
import React from 'react';
import renderer from 'react-test-renderer';
import './__mock__/initMocks';
import { TaxCalculationsComponent } from './TaxCalculations';

jest.mock('./Ui', () => mockComponent);
jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');
jest.mock('@material-ui/core/TableHead', () => 'TableHead');

describe('TaxCalculations', () => {
  it('should render', () => {
    const tree = renderer
      .create(
        <TaxCalculationsComponent
          income={12345678}
          minimumDeduction={12345678}
          personalAllowance={12345678}
          commonIncome={12345678}
          incomeTax={12345678}
          socialSecurityDeduction={12345678}
          step1={12345678}
          step2={12345678}
          step3={12345678}
          step4={12345678}
          step5={12345678}
          totalTax={12345678}
          afterTax={12345678}
          dividends={12345678}
          afterDividendsTax={12345678}
          dividendsTax={12345678}
          surplus={12345678}
          surplusTax={12345678}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
