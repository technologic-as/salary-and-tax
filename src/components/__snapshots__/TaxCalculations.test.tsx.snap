// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaxCalculations should render 1`] = `
<Section
  expanded={true}
  header="Tax calculations"
>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <Table>
        <TableHead>
          <TaxRow
            description="Common income"
          />
        </TableHead>
        <TableBody>
          <TaxRow
            description="Income"
            sum="$12,345,678.00"
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Minimum deduction"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Personal allowance"
            minus={true}
          />
          <TaxRow
            description="Common income"
            sum="$12,345,678.00"
          />
        </TableBody>
      </Table>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <Table>
        <TableHead>
          <TaxRow
            description="Income"
          />
        </TableHead>
        <TableBody>
          <TaxRow
            description="Income"
            sum="$12,345,678.00"
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Income tax"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Social security deduction"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Tax step 1"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Tax step 2"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Tax step 3"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Tax step 4"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Tax step 5"
            minus={true}
          />
          <TaxRow
            amount="$12,345,678.00"
            description="After tax"
            sum="$12,345,678.00"
          />
        </TableBody>
      </Table>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <Table>
        <TableHead>
          <TaxRow
            description="Surplus"
          />
        </TableHead>
        <TableBody>
          <TaxRow
            description="Surplus amount"
            sum="$12,345,678.00"
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Surplus tax"
            minus={true}
            sum="$12,345,678.00"
          />
          <TaxRow
            amount="$12,345,678.00"
            description="Dividends tax"
            minus={true}
            sum="$12,345,678.00"
          />
          <TaxRow
            description="Dividends"
            sum="$12,345,678.00"
          />
        </TableBody>
      </Table>
    </div>
  </div>
</Section>
`;
